File: FORESTGA.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl "\n"

long long n, W, L;
vector<long long> h, r;

bool solve(long long x) {
	long long temp = W;
	for (int i = 0; i < n; i++) {
		long long ht = h[i] + r[i] * x;
		if (ht >= L) {
			temp -= ht;
			if (temp <= 0)
				return true;
		}
	}
	return false;
}

int main(void) {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

#ifndef ONLINE_JUDGE
	freopen("input.txt", "r", stdin);
//	freopen("output.txt", "w", stdout);
#endif

	cin >> n >> W >> L;
	h.resize(n);
	r.resize(n);
	for (int i = 0; i < n; i++)
		cin >> h[i] >> r[i];
	long long l = 0, r = 1e18, ans = LONG_MAX;
	while (l <= r) {
		long long m = l + (r - l) / 2;
		if (solve(m)) {
			ans = min(ans, m);
			r = m - 1;
		} else {
			l = m + 1;
		}
	}
	cout << ans << endl;
}

File: FORESTGA.py
def solve(x):
	temp = w
	for h, r in ls:
		ht = h + r * x
		if ht >= l:
			temp -= ht
			if temp <= 0:
				return True
	return False

n, w, l = map(int, input().split())
ls = []
for _ in range(n):
	h, r = map(int, input().split())
	ls.append((h, r))
L, R = -1, 10 ** 18 + 1
while R - L > 1:
	M = L + (R - L) // 2
	if solve(M):
		R = M
	else:
		L = M
print(R)
